Spotify Liked Songs Queue Skipping

Ben2356 opened the pull request:
When playing the Liked Songs playlist, the user is not able to skip to a specific song in the queue as upon selecting the song it just restarts the first song in the Liked Songs playlist. To get around this, I created a hacky solution that upon selecting the Liked Songs playlist a private dummy playlist will be created through the Spotify Web API (if the user has authenticated) and then populated with all of the songs of the Liked Songs playlist. The cover art and title will remain the same for Liked Songs and from the outside the Liked Songs dummy playlist will look no different than the actual playlist it's mimicking, but this will allow the user to skip to any songs that they wish in the playlist. This dummy playlist will have its state cached in the shared preferences, only updating if the user has changed the Liked Songs playlist content, making it fast upon subsequent loading operations.

This should address this issue: https://github.com/hufman/AndroidAutoIdrive/issues/166

hufman commented on app/src/main/java/me/hufman/androidautoidrive/phoneui/SpotifyAuthorizationActivity.kt:
```diff
@@ -60,7 +60,9 @@ class SpotifyAuthorizationActivity: Activity() {
 			SpotifyScope.USER_MODIFY_PLAYBACK_STATE.uri,
-			SpotifyScope.USER_LIBRARY_READ.uri)
+			SpotifyScope.USER_LIBRARY_READ.uri,
+			SpotifyScope.PLAYLIST_MODIFY_PUBLIC.uri,
```
Maybe the PLAYLIST_MODIFY_PUBLIC permission isn't needed

Ben2356 replied:
Yeah it isn't, I accidentally left this in from when I had the playlists created as public in my initial implementation

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -12,15 +12,19 @@ import com.spotify.android.appremote.api.SpotifyAppRemote
 import com.spotify.protocol.client.Subscription
 import com.spotify.protocol.types.*
 import kotlinx.coroutines.*
+import kotlinx.serialization.decodeFromString
```
The app already has Gson included, which can do class de/serialization, what does the Kotlin version do differently?

Ben2356 replied:
Originally I had used the Kotlinx serialization due to my initial implementation dealing with default parameters for the LikedSongsState and Gson didn't play nicely with those for some reason but now with the latest implementation there really is no need for kotlinx serialization anymore and Gson will do nicely, updated the implementation to use Gson and removed kotlinx serialization

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -244,19 +248,80 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+
+				queueUri = likedSongsState.playlistUri
+
+				if (likedSongsState.queueCoverArtUri == null) {
```
It appears that this block is duplicating `loadQueueCoverart` just to persist the `LikedSongsState.queueCoverArtUri`.
I'm not sure that the LikedSongsState's queueCoverArtUri is strictly necessary: The QueueMetadata.coverArt will start out null either way, so the regular loadQueueCoverArt can load the recently-played Liked Songs coverart and update the QueueMetadata like normal? Caching the previous queueCoverArtUri might be incorrect anyways, I suspect the uri changes to show a different picture when the playlist contents change.

Ben2356 replied:
The scenario around this is a bit interesting as yes the queueCoverArtUri isn't necessary but only in the ideal flow where the user selects the actual Liked Songs playlist. This logic here helps in two ways: 
1. The caching of the cover art uri for the Liked Songs playlist is fast as we don't need to make the extra call to the recently played content API from the Remote API, thus lowering the overall time before the playlist is fully loaded. This is especially useful here as the image uri is always the same static image 
2. If the user selects the dummy Liked Songs playlist, they will then get the cover art that comes from that which is just the 4 cover art images from the first 4 tracks in the playlist by default. We don't want that as we want a single unified experience whether it be the dummy Liked Songs playlist or the real Liked Songs playlist as they should be synonymous and the user shouldn't see a disconnect in that as it might create confusion.

Also the uri for the Liked Songs playlist (the purple to blue gradient background with the white heart image) is always the same for Liked Songs as it is a user collection and I don't believe it is even possible to change it to something else so once that is loaded the initial time that uri will always be correct regardless of the playlist content.

hufman replied:
Ah good point! Thanks for the explanation.

hufman replied:
Thinking on this some more: While it would be helpful for filter out the temporary playlist from the browse list so that the user only selects the normal playlist, this flow is still needed to handle the situation where the context has switched to the temporary playlist and then the car app disconnects and reconnects later, still in this temporary context. And so, I view likedSongsState instead as a persistent tempPlaylistState, remembering the original context's title and coverart and the temp playlist identifier for skipping and matching purposes.

I recommend splitting `createLikedSongsQueueMetadata` in two functions: Reinflating the likedSongsState back into a QueueMetadata object (basically these two else blocks here, which will also be needed for temporary Artist Playlists), and the rest of this function, which clones the Liked Songs (well, any current or given playlist really) to the temporary playlist and persists the temp playlist url, name, and coverart for later reinflation. This will remove the need to duplicate the cover art logic, because it'll just persist the currently-loaded cover art (either Liked or Artist) instead of fetching it again here.

Ben2356 replied:
Ah that's true, I updated the LikedSongsState to be a more generic TemporaryPlaylistState which will allow for greater flexibility in the future when expanding to support Artist playlists. I also agree that the createLikedSongsQueueMetadata() method needed to be split up a bit so I refactored it into the two methods liked you described: one for the creation of a TemporaryPlaylistState and one for the deserialization/inflation of a TemporaryPlaylistState from its JSON representation. I also updated the loadQueueCoverart(..) method so now we can pass in a TemporaryPlaylistState to be updated (and its AppSettings.KEYS key for the save operation) as optional parameters which eliminates the near duplicate code for the fetching of the queue coverart and we can now just keep that logic in one function

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -244,19 +248,80 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+					val uri = webApi.createPlaylist(SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME)
+					if (uri == null) {
+						Log.e(TAG, "Error creating liked songs playlist, falling back to app remote API")
+						queueItems = emptyList()
```
There should be a comment here indicating that we return an empty list to prevent trying the broken skipToIndex behavior in https://github.com/spotify/android-sdk/issues/267

Ben2356 replied:
With this the reasoning behind the emptyList is actually not to prevent the creation of the Liked Songs playlist with any data as we still want this broken behavior since at the very least the user can still access their Liked Songs playlist even if the skipToIndex logic doesn't work as it should. Instead this is so that the createQueueMetadata will properly create the Liked Songs playlist as this conditional will fail otherwise https://github.com/Ben2356/AndroidAutoIdrive/blob/spotifyLikedSongsSkip/app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt#L345

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -223,7 +227,7 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
 
 				val isLikedSongsPlaylist = playerContext.type == "your_library" || playerContext.type == "your_library_tracks"
-				if (isLikedSongsPlaylist) {
+				if (isLikedSongsPlaylist || playerContext.title == SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME) {
```
I think the `createLikedSongsQueueMetadata` isn't needed once Spotify has switched to playing the private copy of the playlist. There might be the possibility of adding new songs to the public Liked Songs that this function could sync, but I'm not sure the real Spotify even supports this, and you have to reclick Liked Songs to play the new song.
If it is needed here, it's probably better to compare based on LikedSongsState.playlistUri (perhaps memoized) because the translation might change.

Ben2356 replied:
This is kind of based on this comment I made above (https://github.com/hufman/AndroidAutoIdrive/pull/263#discussion_r622697262) but the idea here is that the if the user tries to load in the dummy Liked Songs playlist it still needs to be handled as though it were the actual Liked Songs playlist in order to promote the sense of unity between the two and not give the user the idea they are two separate entities. If the user did modify the actual Liked Songs playlist then at this point the private copy would be synced during the createLikedSongsQueueMetadata() method which will avoid desync issues between the two as I think we want to have those as much as sync as possible due to the previously mentioned concept. 

I had thought about comparing it based on a dummy playlist uri member variable that would be set from the createLikedSongsQueueMetadata() otherwise if it's null then loading the LikedSongsState.playlistUri from the AppSettings but with that then the Json has to be parsed out if it's not empty and suddenly a good chunk of the work of createLikedSongsQueueMetadata() is being done every time the queueUri changes and a new QueueMetadata needs to be created for that if condition. Then I was thinking the SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME is not something that would be translated or changed (it shouldn't be) as it's a constant https://github.com/Ben2356/AndroidAutoIdrive/blob/spotifyLikedSongsSkip/app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt#L39 .

I'm actually not sure on this one since I can see what you mean that the playlistUri is a safer way of determining if the user is opening the dummy Liked Songs playlist but at the same time I'm worried about adding that extra performance overhead as it's something that shouldn't be translated or changed otherwise the user risks breaking it.

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -244,19 +248,80 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
 			if (queueItems.isNotEmpty()) {
-				queueMetadata = QueueMetadata("Liked Songs", null, queueItems)
-				loadQueueCoverart()
+				queueMetadata = QueueMetadata(L.MUSIC_LIKED_SONGS_PLAYLIST_NAME, null, queueItems)
```
Good catch, that this should've been translated before, but now that I'm seeing it, why not just use the Spotify-provided name for this playlist that we are overriding? Perhaps save the actual playerContext queueTitle at the top with queueUri and friends, and refer back to it as needed instead of providing our own?

Ben2356 replied:
Ah that's a much better approach than what I was doing here, then we don't have to worry about translations and all that! Added a queueTitle member var and updated it with the playerContext from the callback

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -303,7 +368,7 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
 			if (item != null) {
 				remote.imagesApi.getImage(item.imageUri, Image.Dimension.THUMBNAIL).setResultCallback { coverArt ->
-					queueMetadata = QueueMetadata(item.title, item.subtitle, queueItems, coverArt, item.uri)
+					queueMetadata?.coverArt = coverArt
```
What does this change improve? Are there places in the code that hold a reference to the queueMetadata that need the new coverart? If it's just to reuse the fake list of queueItems, it can probably be fetched from `QueueMetadata(.., .., queueMetadata.queueItems, ..)`

Ben2356 replied:
The reasoning for this change was I was looking at this more in depth as part of this task and I noticed we are creating a brand new QueueMetadata object just to introduce a cover art bitmap not only adds extra overhead in creating a new instance of the queueMetadata (yes it's a very small maybe immeasurable impact 😄) but also seems unnecessary and not as clean as simply updating the cover art on the existing queueMetadata instance. This way we only have a single queueMetadata instance (during the queue "session") that is being updated as the data is becoming available for it

hufman replied:
Good thinking, for considering performance on phones! However, the new queueMetadata objects should be relatively rare, though I do intend to run a real performance analysis to know for certain where any hot spots are.
Some places in the code (such as `EnqueuedView.redraw()`) hold a reference to the previous version of queueMetadata, checking object equality to know when the controller has a new queueMetadata to show. Simply updating the existing object won't notify the change properly.

Ben2356 replied:
Ah I see what you mean and digging further into Kotlin data classes it turns out that they automatically implement the equals(..) method so all of those references to the QueueMetadata won't be doing object equality checks on the address of the object but will do member variable equality which then means that we won't have to worry about those equality checks not returning correctly for those scenarios (docs for reference: https://kotlinlang.org/docs/data-classes.html)

hufman replied:
This is true, indeed! However, the EnqueuedView holds a reference to SpotifyAppController's queueMetadata, and so updating this queueMetadata object will also mean that the EnqueuedView's queueMetadata will immediately have this new coverart, which means its redraw logic won't notice the update and any deferred coverart loading won't display until the next time the EnqueuedView is opened.
I'd prefer to leave the QueueMetadata class as immutable to reduce the possible side effects.

Ben2356 replied:
Ohh I think I'm understanding now, I agree it is better to leave the QueueMetadata class as immutable then. I reverted the QueueMetadata to be immutable and reverted this logic

Ben2356 replied:
It looks like my comment was deleted somehow, I think when my session glitched and I went to remove the extra comment I accidentally deleted this one as well. 

I reverted the QueueMetadata modifications and the code to set the coverArt to how it was before

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -244,19 +248,80 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+				val hashCode = queueItems.hashCode().toString()
+				val likedSongsStateJson = appSettings[AppSettings.KEYS.SPOTIFY_LIKED_SONGS_PLAYLIST_STATE]
+				val likedSongsState: LikedSongsState
+				if (likedSongsStateJson.isBlank()) {
```
This will get blanked out when the user Clears Storage. Can the WebApi offer a way to discover a previously-created AndroidAutoIdrive playlist instead of always creating? Maybe: !public, has the same queueTitle as the real `your_library` playlist, but is not the `your_library` playlist.

hufman replied:
I don't want to rely on string matching if possible because the translations are mutable.

Ben2356 replied:
Ah that's a good idea to have a discovery of an already existing dummy playlist before creating a new one. Unfortunately it seems as though the only possible way to find the appropriate playlist will have to be through string matching either the name but as I mentioned in this comment here:  https://github.com/hufman/AndroidAutoIdrive/pull/263#discussion_r622723818 the name of the Liked Songs dummy playlist (SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME) will be a constant that should not be translated or tampered with so we can use this playlist title to find the appropriate playlist. The problem with the queueTitle approach you suggested is that the queueTitle will be translated and thus needs to be stored as to figure out what the dummy playlist is but then since the storage is being cleared we lose that and now we don't know what to look for. 

I added a new conditional check that if the likedSongsStateJson is blank then it will first check the user's playlists to see if there is an already existing dummy Liked Songs playlist. If that playlist is found it will pull the playlist uri and use that instead of creating a new one

hufman replied:
Ah, I had assumed the `SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME` was more visible, and thus would prefer to be translated. By having the translation being just the description, that does allow the name to be used as an identifier.

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt:
```diff
@@ -72,6 +76,79 @@ class SpotifyWebApi private constructor(val context: Context, val appSettings: M
+	 * Creates a private playlist with the provided name. The newly created playlist's [PlaylistUri]
+	 * is returned.
+	 */
+	suspend fun createPlaylist(playlistName: String): PlaylistUri? {
```
Should the description be passed in as well, allowing for different descriptions for different playlists in the future?

Ben2356 replied:
Yeah it should be to ensure greater flexibility for future expansion, updated this method signature to take in a playlist description parameter that defaults to null

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt:
```diff
@@ -72,6 +76,79 @@ class SpotifyWebApi private constructor(val context: Context, val appSettings: M
+	 */
+	suspend fun getPlaylistUri(playlistName: String): SpotifyUri? {
+		try {
+			val playlists = webApi?.playlists?.getClientPlaylists(15,0)
```
Providing a limit of 15 seems too small, I'd be fine setting this to the max of 50, and since it's an async function, it might be feasible to implement pagination easily enough.

Ben2356 replied:
Agreed, updated the limit to 50 and also added pagination support for the getClientPlaylists built in pagination system

hufman commented on app/src/main/res/values/strings.xml:
```diff
@@ -289,6 +289,7 @@
     <string name="MUSIC_TURN_REPEAT_ALL_ON">Turn Repeat All On</string>
     <string name="MUSIC_TURN_REPEAT_ONE_ON">Turn Repeat One On</string>
     <string name="MUSIC_TURN_REPEAT_OFF">Turn Repeat Off</string>
+    <string name="MUSIC_TEMPORARY_PLAYLIST_DESCRIPTION">AndroidAutoIDrive dummy playlist - please do not delete</string>
```
I'd prefer a wording like "AndroidAutoIDrive temporary playlist", and I'm not sure the "please do not delete" is strictly necessary: To me, it gives the impression that Bad Things Will Happen if it is deleted, as opposed to a slight de-optimization path. I think the effort to recreate it is only slightly more than the effort to replace the contents with any updated Liked Songs contents.

Ben2356 replied:
Updated the wording on that and removed the "please do not delete" part. You're right, it isn't really necessary and if the user does delete it recreation really isn't that intense of an operation. I haven't measured the difference in the time it takes to replace the contents of a temporary playlist in comparison to the full creation of a playlist but I think you're right that they would be very similar. I should eventually run some performance metrics and see how all of the possible flows compare, mostly out of curiosity 😃

hufman commented on app/src/main/java/me/hufman/androidautoidrive/L.kt:
```diff
@@ -53,6 +53,7 @@ object L {
 	var MUSIC_TURN_REPEAT_ALL_ON = "Turn Repeat All On"
 	var MUSIC_TURN_REPEAT_ONE_ON = "Turn Repeat One On"
 	var MUSIC_TURN_REPEAT_OFF = "Turn Repeat Off"
+	var MUSIC_TEMPORARY_PLAYLIST_DESCRIPTION = "AndroidAutoIDrive dummy playlist - please do not delete"
```
Could this match the strings.xml value, just for cleanliness? The only place it would possibly be seen is in unit tests, but just for cleanliness' sake!

Ben2356 replied:
Yeah definitely, sorry I missed changing that string. Updated to match the strings.xml equivalent key

hufman commented on app/src/main/res/values/strings.xml:
```diff
@@ -373,6 +373,7 @@
     <string name="MUSIC_TURN_REPEAT_ALL_ON">Turn Repeat All On</string>
     <string name="MUSIC_TURN_REPEAT_ONE_ON">Turn Repeat One On</string>
     <string name="MUSIC_TURN_REPEAT_OFF">Turn Repeat Off</string>
+    <string name="MUSIC_TEMPORARY_PLAYLIST_DESCRIPTION">AndroidAutoIDrive temporary playlist</string>
```
I'm planning on renaming the project soon as part of 1.3.1, can this be pre-emptively updated to AAIdrive?

Ben2356 replied:
Yeah sure thing, updated this as well as the corresponding string from L.kt

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt:
```diff
@@ -34,6 +37,7 @@ class SpotifyWebApi private constructor(val context: Context, val appSettings: M
 		const val TAG = "SpotifyWebApi"
 		const val NOTIFICATION_CHANNEL_ID = "SpotifyAuthorization"
 		const val NOTIFICATION_REQ_ID = 56
+		const val LIKED_SONGS_PLAYLIST_NAME = "ANDROIDAUTOIDRIVE_LIKED_SONGS"
```
I'm planning to rename the project soon for 1.3.1, can this be preemptively changed to `AAIDRIVE_LIKED_SONGS`?

Ben2356 replied:
Yeah, updated to "AAIDRIVE_LIKED_SONGS"

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -242,19 +247,106 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+			temporaryPlaylistState.hashCode = queueItemsHashCode
+		}
+
+		queueUri = temporaryPlaylistState.playlistUri
```
I believe this is unused

Ben2356 replied:
This assignment is needed here as the queueUri state variable needs to be updated with the current session's queueUri that is related to the temporary playlist after it is either created or loaded since this will be different than the actual queueUri of the user library/playlist we are creating the temporary playlist of (this should stay in sync with the our currently playing playlist since we are essentially playing the temporary playlist and not the actual Liked Songs playlist for example)

hufman replied:
Oh oh! Right.

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -242,19 +247,106 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+
+		val temporaryPlaylistState = TemporaryPlaylistState(queueItemsHashCode, playlistUri, playlistId, null)
+
+		queueUri = temporaryPlaylistState.playlistUri
```
I believe this is unused

Ben2356 replied:
Please see this comment: https://github.com/hufman/AndroidAutoIdrive/pull/263#discussion_r690780026

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -292,16 +384,22 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+	 * with the result.
 	 */
-	private fun loadQueueCoverart() {
+	private fun loadQueueCoverart(temporaryPlaylistState: TemporaryPlaylistState? = null, appSettingsTemporaryPlaylistStateKey: AppSettings.KEYS? = null) {
```
I don't like this code smell, which signals that this function is doing too much work: I believe the extra complication is to save the current queue's coverArtUri to the TemporaryPlaylistState. I wonder if this extra retention is actually necessary, however.
When the user plays the Liked Songs playlist, this loadQueueCoverart will load the proper heart banner normally. After the user clicks a queue entry and the context switches to the temporary playlist, this function will return the normal Spotify coverart for the temporary playlist, which might be fine in my estimation. Alternatively, the recentlyPlayed results could skip the TemporaryPlaylist if it's the first result and assume that the next item is the real Liked Songs.
However if the TemporaryPlaylistState really wants to hold its source coverArtUri, I'd prefer to have `createTemporaryPlaylistState` always do the regular `loadQueueCoverart` first (additionally allowing Car UI to show the queue cover art during any playlist synchronization) and then create and cache the entire coherent object using the real queueCoverArt information, instead of trying to recover a missing coverArtUri later in `loadTemporaryPlaylistState`. Such guessing will be even harder when Artist queues get added. The calculated queueCoverArtUri can go live as an additional member at the top with the queueTitle for the wrapping `createTemporaryPlaylistState` to use.

hufman commented:
Can the Browse listing omit the temporary playlist from the Playlist results, perhaps?

Ben2356 commented:
> 
> 
> Can the Browse listing omit the temporary playlist from the Playlist results, perhaps?

Yeah it can and after giving it some thought I should have done this originally as to not confuse the user with such an oddly named playlist. I added this logic to remove it from the browse listing and also added a unit test for this :)

hufman commented on app/build.gradle:
```diff
@@ -398,6 +398,8 @@ dependencies {
     implementation ('com.adamratzman:spotify-api-kotlin-android:3.7.0') {
         exclude group: 'org.jetbrains.kotlinx', module: 'kotlinx-coroutines-core'
     }
+    implementation "com.soywiz.korlibs.korim:korim-android:2.2.0"
```
I suspect this was a previous direction, but seems no longer used presently

Ben2356 replied:
This is kind of an interesting scenario since when I was trying to utilize the uploadClientPlaylistCover from the ClientPlaylistAPI I was running into the issue that it was failing to compile and also experiencing runtime errors due to the korim and korio dependencies not being found. I looked into this issue on the SpotifyWebApi library's discord server and turns out there is a bug/something broken around this and the developer has been busy lately so this is kind of a hacky workaround in the meantime until that fix is released 😅

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -311,10 +317,13 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+					queueMetadata = QueueMetadata(queueTitle, item.subtitle, queueItems, coverArt, item.uri)
+
+					GlobalScope.launch(defaultDispatcher) {
+						val coverArtBase64 = Base64.encodeToString(Utils.compressBitmapJpg(coverArt, 100), Base64.NO_WRAP)
```
A quality of ~85 is acceptable and saves a bit of space

Ben2356 replied:
Good point, updated the quality to 85

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -242,17 +249,115 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+
+		queueUri = temporaryPlaylistState.playlistUri
+
+		val recentlyPlayedUri = "com.spotify.recently-played"
```
This duplicated piece of code could be combined with the regular `loadQueueCoverart` if the `webApi.setPlaylistImage` was conditional (perhaps `playerContext.title != SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME`).
Alternatively, `loadQueueCoverart` could be converted to a suspend function with a CompletableDeferred, similar to how the [adapter to the native browse functionality works](https://github.com/hufman/AndroidAutoIdrive/blob/e5968c276b587537e12f791444386fd82ff3dbbf/app/src/main/java/me/hufman/androidautoidrive/music/MusicBrowser.kt#L123), and return the coverart instead of replacing the queueMetadata.

Ben2356 replied:
Ah I like the concept of the loadQueueCoverart just returning a Bitmap as a suspend function with a CompletableDeferred, that makes it more generic and allows the consumers to define what to do with that coverArt image which is better in my opinion than trying to cram too much logic inside for all the different scenarios. I updated the function and renamed it now as ```getQueueCoverArt()``` and updated the logic throughout that call this method

Ben2356 commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:
```diff
@@ -292,7 +397,7 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
 	}
 
 	/**
-	 * Creates a new instance of the [QueueMetadata] with the queue cover art loaded.
```
As per my change to this function mentioned in the above comment (https://github.com/hufman/AndroidAutoIdrive/pull/263#discussion_r695048256) I replaced this doc with a more appropriate description for ```getQueueCoverArt()```

Ben2356 commented:
> 
> 
> I like it! It works pretty well, and the switchover from the Liked Songs playlist to the AAIDRIVE_LIKED_SONGS playlist is very seamless, and I'm delighted that the current queue item checkmark is correctly positioned before and after the cutover.
> 
> I had one hiccup when first changing to this build: The SpotifyWebAPI was encountering an error with the narrower scopes from my previous auth session, could you detect this and clear the cookie to reprompt for auth with the new scopes?
> `E/SpotifyWebApi: Exception occurred when trying to create playlist AAIDRIVE_LIKED_SONGS with the message: Received Status Code 403. Error cause: Insufficient client scope. Reason: URL: https://api.spotify.com/v1/users/.../playlists`

That's awesome, I'm very glad to hear that 😃! I added in two new improvements for addressing that issue, the first of which is a new wrapper method around the calls to the SpotifyWebAPI library that will worry about exception handling so we don't have so much duplicated code around each method and also added a scenario to catch if there is a 403 Status Code exception and re-prompt the user to reauthenticate since they are likely using an invalid token. This should fix that issue :)

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt:
```diff
@@ -312,6 +249,41 @@ class SpotifyWebApi private constructor(val context: Context, val appSettings: M
+	 *
+	 * Note: This should wrap **EVERY** [SpotifyClientApi] call.
+	 */
+	private suspend fun <R> executeApiCall(errorMessage: String, block: suspend () -> R): R? {
```
Oooo I like this a lot!
